### PR TITLE
DGVA-64 View list of studies

### PR DESCRIPTION
**Pull request by tomdcsmith**
- New dgva-server package for dgva webservices
- dgva-server using DGVAPRO
- Listing structural studies moved from eva-server to dgva-server, using dgva study browser in DGVAPRO

### CODE REVIEW

File: dgva-server/src/main/java/uk/ac/ebi/dgva/server/security/CORSResponseFilter.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletResponse;
+
+import org.springframework.stereotype.Component;
+
+/**
+ *
+ * @author Cristina Yenyxe Gonzalez Garcia <cyenyxe@ebi.ac.uk>
```
Comment (jmmut) [inline dgva-server/src/main/java/uk/ac/ebi/dgva/server/security/CORSResponseFilter.java]: remove author tag

File: eva-lib/src/main/java/uk/ac/ebi/eva/lib/utils/QueryUtils.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+import javax.servlet.http.HttpServletRequest;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+@Component
+public class QueryUtils {
+
+    protected final String version = "v1";
```
Comment (jmmut) [inline eva-lib/src/main/java/uk/ac/ebi/eva/lib/utils/QueryUtils.java]: can both dgva and eva have different versions? they still have a duplicated attribute for it

File: eva-lib/src/main/java/uk/ac/ebi/eva/lib/utils/QueryUtils.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    @Autowired
+    protected HttpServletRequest httpServletRequest;
+
+    private QueryOptions queryOptions;
+    protected long startTime;
+    protected long endTime;
+
+    protected static Logger logger = LoggerFactory.getLogger(QueryUtils.class);
+
+    public QueryOptions getQueryOptions() {
```
Comment (jmmut) [inline eva-lib/src/main/java/uk/ac/ebi/eva/lib/utils/QueryUtils.java]: does this need to be public? what advantage does this method provide in contrast with a protected attribute? are the startTime and endTime required to be protected still?

File: dgva-server/src/main/java/uk/ac/ebi/dgva/server/Profiles.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.ac.ebi.dgva.server;
+
+/**
+ * Created by jorizci on 01/11/16.
```
Comment (jmmut) [inline dgva-server/src/main/java/uk/ac/ebi/dgva/server/Profiles.java]: remove

File: dgva-server/src/main/java/uk/ac/ebi/dgva/server/configuration/Swagger2Configuration.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                .apiInfo(apiInfo())
+                .pathMapping("/")
+                .select()
+                .apis(RequestHandlerSelectors.basePackage("uk.ac.ebi.dgva.server"))
+                .build();
+    }
+
+    private ApiInfo apiInfo() {
+        return new ApiInfoBuilder()
+                .title("European Variation Archive REST Web Services API")
```
Comment (jmmut) [inline dgva-server/src/main/java/uk/ac/ebi/dgva/server/configuration/Swagger2Configuration.java]: should this be the same? I don't know all the context of dgva, that's why I ask if it should be reflected here

File: dgva-server/src/main/java/uk/ac/ebi/dgva/server/ws/ArchiveWSServer.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+import org.springframework.web.bind.annotation.RestController;
+
+import uk.ac.ebi.eva.lib.metadata.StudyDgvaDBAdaptor;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Properties;
+
+@RestController
+@RequestMapping(value = "/v1/meta", produces = "application/json")
```
Comment (jmmut) [inline dgva-server/src/main/java/uk/ac/ebi/dgva/server/ws/ArchiveWSServer.java]: I got a hunch that this hardcoded version will bite us this year. Likelihood, @cyenyxe ?

File: eva-lib/src/main/java/uk/ac/ebi/eva/lib/utils/QueryUtils.java (hunk 3 above)
Comment (tomdcsmith) [inline eva-lib/src/main/java/uk/ac/ebi/eva/lib/utils/QueryUtils.java]: This method is used outside of this package, inside EvaWSServer and DgvaWSServer in the eva-server and dgva-server packages.

File: eva-lib/src/main/java/uk/ac/ebi/eva/lib/utils/QueryUtils.java (hunk 2 above)
Comment (tomdcsmith) [inline eva-lib/src/main/java/uk/ac/ebi/eva/lib/utils/QueryUtils.java]: @cyenyxe should the versions be different between the two?

File: dgva-server/src/main/java/uk/ac/ebi/dgva/server/configuration/Swagger2Configuration.java (hunk 5 above)
Comment (tomdcsmith) [inline dgva-server/src/main/java/uk/ac/ebi/dgva/server/configuration/Swagger2Configuration.java]: I didn't change this since I assumed the goal was to merge EVA and DGVa as much as possible. @cyenyxe, shall I change this to:

```
return new ApiInfoBuilder()
                .title("Database  REST Web Services API")
                .contact(new Contact("Database of Genomic Variants Archive team", "http://www.ebi.ac.uk/dgva", "dgva-helpdesk@ebi.ac.uk

"))
                .license("Apache License Version 2.0")
                .licenseUrl("https://www.apache.org/licenses/LICENSE-2.0")
                .version("1.0")
                .build();
```

File: eva-lib/src/main/java/uk/ac/ebi/eva/lib/utils/QueryUtils.java (hunk 2 above)
Comment (cyenyxe) [inline eva-lib/src/main/java/uk/ac/ebi/eva/lib/utils/QueryUtils.java]: Yes, it could be possible when we start updating one of the APIs first. That work will be first focused on EVA's.

File: eva-lib/src/test/java/uk/ac/ebi/eva/lib/metadata/DgvaStudyTestData.java [hunk 7]
```diff
@@ -36,16 +36,16 @@
...
     static void persistTestData(TestEntityManager entityManager) {
-        DgvaStudyBrowser study1 = new DgvaStudyBrowser(STUDY_1_ID, 1000, 10, 100, "9606", HUMAN, HOMO_SAPIENS, "1111",
-                                                       "alias1", "Study 1", CONTROL_SET, "PRJ1", "www.study1.com",
+        DgvaStudyBrowser study1 = new DgvaStudyBrowser(STUDY_1_ID, "9606", HUMAN, HOMO_SAPIENS, "1111",
+                                                       "Study 1", CONTROL_SET, "PRJ1", "www.study1.com",
                                                        "This is study 1", "Sequence alignment", "HMM", "Sequencing",
                                                        "Illumina", "GRCh38");
-        DgvaStudyBrowser study2= new DgvaStudyBrowser("estd2", 2000, 20, 200, "9606", HUMAN, HOMO_SAPIENS, "2222",
-                                                      "alias2", "Study 2", COLLECTION, "PRJ2", "www.study2.com",
+        DgvaStudyBrowser study2= new DgvaStudyBrowser("estd2", "9606", HUMAN, HOMO_SAPIENS, "2222",
```
Comment (cyenyxe) [inline eva-lib/src/test/java/uk/ac/ebi/eva/lib/metadata/DgvaStudyTestData.java]: Missing space before the equals sign, requires block formatting.

File: eva-server/src/main/java/uk/ac/ebi/eva/server/configuration/UtilsConfiguration.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+ */
+
+package uk.ac.ebi.eva.server.configuration;
+
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@ComponentScan(basePackages = {"uk.ac.ebi.eva.lib.utils"})
+public class UtilsConfiguration {
```
Comment (cyenyxe) [inline eva-server/src/main/java/uk/ac/ebi/eva/server/configuration/UtilsConfiguration.java]: Why not making the application scan "uk.ac.ebi.eva.lib" instead? Otherwise we will need a configuration class for each sub-package.

File: eva-server/src/main/java/uk/ac/ebi/eva/server/ws/ArchiveWSServer.java [hunk 9]
```diff
@@ -127,20 +119,20 @@ public QueryResponse getStudiesStats(@RequestParam(name = "species", required =
                                          @RequestParam(name = "structural", defaultValue = "false") boolean structural) {
         initializeQuery();
         if (species != null && !species.isEmpty()) {
-            queryOptions.put("species", species);
+            getQueryOptions().put("species", species);
```
Comment (cyenyxe) [inline eva-server/src/main/java/uk/ac/ebi/eva/server/ws/ArchiveWSServer.java]: The first call could assign them to a local variable instead of calling the method multiple times.

File: eva-lib/src/main/java/uk/ac/ebi/eva/lib/models/VariantStudy.java [hunk 10]
```diff
@@ -116,14 +115,15 @@ public VariantStudy(String studyName, String studyId) {
...
 
-    public VariantStudy(String studyName, String studyId, List<VariantSource> sources, String description, int[] taxonomyId, 
-            String speciesCommonName, String speciesScientificName, String sourceType, String center, String material, 
-            String scope, StudyType type, String experimentType, String experimentTypeAbbreviation, String referenceAssembly, 
-            String platform, URI projectUrl, String[] publications, int numVariants, int numSamples) {
+    public VariantStudy(String studyName, String studyId, List<VariantSource> sources, String description,
+                        int[] taxonomyId, String speciesCommonName, String speciesScientificName, String sourceType,
+                        String center, String material, String scope, StudyType type, String experimentType,
+                        String experimentTypeAbbreviation, String referenceAssembly, String platform, URI projectUrl,
+                        String[] publications, int numSamples) {
```
Comment (cyenyxe) [inline eva-lib/src/main/java/uk/ac/ebi/eva/lib/models/VariantStudy.java]: Didn't we find that DGVa supports multiple species and assemblies in a single study? Or is that refactoring intended for a future PR? I am fine with splitting it in 2 if that facilitates merging.

Comment (tomdcsmith) [inline eva-lib/src/main/java/uk/ac/ebi/eva/lib/models/VariantStudy.java]: I think it would be easier to make it two different PRs. That refactoring might need some extra thought and additional testing. It also might be outside the scope of moving the dgva webservice functionality from eva-server to the new dgva-server.

File: eva-server/src/main/java/uk/ac/ebi/eva/server/ws/ArchiveWSServer.java (hunk 9 above)
Comment (tomdcsmith) [inline eva-server/src/main/java/uk/ac/ebi/eva/server/ws/ArchiveWSServer.java]: There are changes to this in the PR https://github.com/EBIvariation/eva-ws/pull/52. No local variable is created, but query utils is used directly.

File: dgva-server/src/main/java/uk/ac/ebi/dgva/server/Profiles.java [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.ac.ebi.dgva.server;
+
+public class Profiles {
```
Comment (tomdcsmith) [inline dgva-server/src/main/java/uk/ac/ebi/dgva/server/Profiles.java]: Since the authentication was removed from dgva-server this isn't used in dgva-server anyway, so I can remove it.

File: dgva-server/src/main/resources/application.properties [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+spring.datasource.url=@dgvapro.host@
+spring.datasource.username=@dgvapro.user@
+spring.datasource.password=@dgvapro.passwd@
+spring.datasource.driver-class-name=oracle.jdbc.OracleDriver
+spring.datasource.type=oracle.jdbc.pool.OracleDataSource
+spring.jpa.database-platform=org.hibernate.dialect.Oracle10gDialect
+
+#Avoid hibernate ddl schema generation/update or validation
+spring.jpa.hibernate.ddl-auto = none
+#spring.jpa.properties.hibernate.dialect = org.hibernate.dialect.OracleDialect
```
Comment (tomdcsmith) [inline dgva-server/src/main/resources/application.properties]: I think this is taken care of in the property higher up

`spring.jpa.database-platform=org.hibernate.dialect.Oracle10gDialect`

I don't think this should be needed, this property isn't used in this example https://www.mkyong.com/spring-boot/spring-boot-spring-data-jpa-oracle-example/

Comment (cyenyxe) [inline dgva-server/src/main/resources/application.properties]: Could it be moved to the bottom for more clear grouping of properties (`spring.datasource` vs `spring.jpa`)? Also, please check if "OracleDialect" (without "10g") also works as it will be more generic and could still work in case of upgrade.

File: dgva-server/src/main/resources/application.properties [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+
+spring.jmx.enabled = false
+
+springfox.documentation.swagger.v2.path=/webservices/api
+
+#spring.profiles.active=oauth2-security
+#security.oauth2.resource.user-info-uri = ...
```
Comment (cyenyxe) [inline dgva-server/src/main/resources/application.properties]: No security = no need to keep these 2 properties.

File: dgva-server/src/main/java/uk/ac/ebi/dgva/server/configuration/Swagger2Configuration.java (hunk 5 above)
Comment (cyenyxe) [inline dgva-server/src/main/java/uk/ac/ebi/dgva/server/configuration/Swagger2Configuration.java]: Confirmed that we must keep using DGVa.

File: dgva-server/src/main/java/uk/ac/ebi/dgva/server/configuration/JpaRepositoryConfiguration.java [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+package uk.ac.ebi.dgva.server.configuration;
+
+import org.springframework.boot.autoconfigure.domain.EntityScan;
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.data.jpa.repository.config.EnableJpaRepositories;
+import uk.ac.ebi.eva.lib.extension.ExtendedJpaRepositoryFunctionsImpl;
+
+@Configuration
+@EnableJpaRepositories(basePackages = {"uk.ac.ebi.eva.lib.repository"},
```
Comment (pabarcgar) [inline dgva-server/src/main/java/uk/ac/ebi/dgva/server/configuration/JpaRepositoryConfiguration.java]: The package name is "...eva.lib.repositories", not "...repository".

File: dgva-server/src/main/java/uk/ac/ebi/dgva/server/ws/ArchiveWSServer.java [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+    private StudyDgvaDBAdaptor studyDgvaDbAdaptor;
+
+    private Properties properties;
+    
+    public ArchiveWSServer() throws IOException {
+    }
+
+    @RequestMapping(value = "/studies/all", method = RequestMethod.GET)
+    public QueryResponse getStudies(@RequestParam(name = "species", required = false) List<String> species,
+                                    @RequestParam(name = "type", required = false) List<String> types) {
```
Comment (pabarcgar) [inline dgva-server/src/main/java/uk/ac/ebi/dgva/server/ws/ArchiveWSServer.java]: Is it the name "type" correct if it can accept a list of types? 

How do Spring expect those parameters?:
 ...?type=type1,type2
or 
...?type=type1&type=type2

File: dgva-server/src/test/java/uk/ac/ebi/dgva/server/ws/ArchiveWSServerTest.java [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+        ResponseEntity<QueryResponse<QueryResult<VariantStudy>>> response = restTemplate.exchange(
+                url, HttpMethod.GET, null,
+                new ParameterizedTypeReference<QueryResponse<QueryResult<VariantStudy>>>() {});
+        assertEquals(HttpStatus.OK, response.getStatusCode());
+
+        QueryResponse<QueryResult<VariantStudy>> queryResponse = response.getBody();
+        assertEquals(1, queryResponse.getResponse().size());
+
+        List<VariantStudy> results = queryResponse.getResponse().get(0).getResult();
+        assertTrue(results.size() >= 1);
```
Comment (pabarcgar) [inline dgva-server/src/test/java/uk/ac/ebi/dgva/server/ws/ArchiveWSServerTest.java]: There are 3 studies. Can we change this to results.size() == 3?

File: eva-lib/src/test/java/uk/ac/ebi/eva/lib/metadata/ArchiveDgvaDBAdaptorTest.java [hunk 17]
```diff
@@ -23,6 +23,7 @@
 import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;
 import org.springframework.test.context.junit4.SpringRunner;
 
+import uk.ac.ebi.eva.lib.metadata.dgva.ArchiveDgvaDBAdaptor;
```
Comment (pabarcgar) [inline eva-lib/src/test/java/uk/ac/ebi/eva/lib/metadata/ArchiveDgvaDBAdaptorTest.java]: This test should be moved to the same package that the class that is testing:  uk.ac.ebi.eva.lib.metadata.dgva.

File: eva-lib/src/test/java/uk/ac/ebi/eva/lib/metadata/ArchiveEvaproDBAdaptorTest.java [hunk 18]
```diff
@@ -25,6 +25,7 @@
 
 import uk.ac.ebi.eva.lib.entities.Project;
 import uk.ac.ebi.eva.lib.entities.Taxonomy;
+import uk.ac.ebi.eva.lib.metadata.eva.ArchiveEvaproDBAdaptor;
```
Comment (pabarcgar) [inline eva-lib/src/test/java/uk/ac/ebi/eva/lib/metadata/ArchiveEvaproDBAdaptorTest.java]: This test should be moved to the same package that the class that is testing:  uk.ac.ebi.eva.lib.metadata.eva

File: eva-lib/src/test/java/uk/ac/ebi/eva/lib/metadata/StudyDgvaDBAdaptorTest.java [hunk 19]
```diff
@@ -23,6 +23,7 @@
 import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;
 import org.springframework.test.context.junit4.SpringRunner;
 
+import uk.ac.ebi.eva.lib.metadata.dgva.StudyDgvaDBAdaptor;
```
Comment (pabarcgar) [inline eva-lib/src/test/java/uk/ac/ebi/eva/lib/metadata/StudyDgvaDBAdaptorTest.java]: This test should be moved to the same package that the class that is testing: uk.ac.ebi.eva.lib.metadata.dgva

File: eva-lib/src/test/java/uk/ac/ebi/eva/lib/metadata/StudyEvaproDBAdaptorTest.java [hunk 20]
```diff
@@ -24,6 +24,7 @@
 import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;
 import org.springframework.test.context.junit4.SpringRunner;
 
+import uk.ac.ebi.eva.lib.metadata.eva.StudyEvaproDBAdaptor;
```
Comment (pabarcgar) [inline eva-lib/src/test/java/uk/ac/ebi/eva/lib/metadata/StudyEvaproDBAdaptorTest.java]: This test should be moved to the same package that the class that is testing: uk.ac.ebi.eva.lib.metadata.eva

File: eva-lib/src/test/java/uk/ac/ebi/eva/lib/metadata/VariantSourceEvaProDBAdaptorTest.java [hunk 21]
```diff
@@ -23,6 +23,7 @@
 import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;
 import org.springframework.test.context.junit4.SpringRunner;
 import uk.ac.ebi.eva.commons.core.models.stats.VariantSourceStats;
+import uk.ac.ebi.eva.lib.metadata.eva.VariantSourceEvaProDBAdaptor;
```
Comment (pabarcgar) [inline eva-lib/src/test/java/uk/ac/ebi/eva/lib/metadata/VariantSourceEvaProDBAdaptorTest.java]: This test should be moved to the same package that the class that is testing: uk.ac.ebi.eva.lib.metadata.eva

File: eva-server/src/main/java/uk/ac/ebi/eva/server/ws/ArchiveWSServer.java [hunk 22]
```diff
@@ -85,21 +83,17 @@ public QueryResponse countStudies() {
 
     @RequestMapping(value = "/studies/all", method = RequestMethod.GET)
     public QueryResponse getStudies(@RequestParam(name = "species", required = false) String species,
-                                    @RequestParam(name = "type", required = false) String types,
-                                    @RequestParam(name = "structural", defaultValue = "false") boolean structural) {
+                                    @RequestParam(name = "type", required = false) String types) {
```
Comment (pabarcgar) [inline eva-server/src/main/java/uk/ac/ebi/eva/server/ws/ArchiveWSServer.java]: The same comment that in the previous server: type or types?

File: dgva-server/pom.xml [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+            <url>https://maven.oracle.com</url>
+            <layout>default</layout>
+            <releases>
+                <enabled>true</enabled>
+                <updatePolicy>always</updatePolicy>
+            </releases>
+        </repository>
+    </repositories>
+
+    <pluginRepositories>
```
Comment (tomdcsmith) [inline dgva-server/pom.xml]: It is according to Oracle's tutorial https://blogs.oracle.com/dev2dev/get-oracle-jdbc-drivers-and-ucp-from-oracle-maven-repository-without-ides

File: dgva-server/src/main/java/uk/ac/ebi/dgva/server/ws/ArchiveWSServer.java (hunk 15 above)
Comment (tomdcsmith) [inline dgva-server/src/main/java/uk/ac/ebi/dgva/server/ws/ArchiveWSServer.java]: The first example, I think. @cyenyxe are you ok with this being changed to "types"?

Comment (cyenyxe) [inline dgva-server/src/main/java/uk/ac/ebi/dgva/server/ws/ArchiveWSServer.java]: Changing it to "types" will break compatibility so let's leave it as is for now.

**Comment by tomdcsmith** [comment]
variation-commons PR https://github.com/EBIvariation/variation-commons/pull/41 needs merging first.
